### PR TITLE
Added caching for ActionConstraints

### PR DESCRIPTION
**Pull request by ajaybhargavb**
@rynowak / @Eilon

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Core/Infrastructure/DefaultActionSelector.cs [hunk 1]
```diff
@@ -33,10 +35,12 @@ public class DefaultActionSelector : IActionSelector
         public DefaultActionSelector(
             IActionSelectorDecisionTreeProvider decisionTreeProvider,
             IEnumerable<IActionConstraintProvider> actionConstraintProviders,
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Infrastructure/DefaultActionSelector.cs]: We shouldn't need the providers here anymore

File: src/Microsoft.AspNet.Mvc.Core/Internal/ActionConstraintCache.cs [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+                    var item = items[i];
+                    if (item.IsReusable)
+                    {
+                        Items[i] = item;
+                    }
+                    else
+                    {
+                        Items[i] = new ActionConstraintItem(item.Metadata);
+                    }
+                }
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Internal/ActionConstraintCache.cs]: It looks like this is an extra defensive copy of `items` - the filter cache code may have had this too. Can you take a look and tell me if you agree?

**Comment by rynowak** [comment]
:watch: for small comments

File: src/Microsoft.AspNet.Mvc.Core/Internal/ActionConstraintCache.cs (hunk 2 above)
Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.Core/Internal/ActionConstraintCache.cs]: I agree. We do the same thing here https://github.com/aspnet/Mvc/pull/3965/files#diff-450e0a08195e9893206a2926f8b38335R106

**Comment by ajaybhargavb** [comment]
:up: :date:

File: src/Microsoft.AspNet.Mvc.Core/Internal/ActionConstraintCache.cs [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.AspNet.Http;
+using Microsoft.AspNet.Mvc.Abstractions;
+using Microsoft.AspNet.Mvc.ActionConstraints;
+using Microsoft.AspNet.Mvc.Infrastructure;
+
+namespace Microsoft.AspNet.Mvc.Internal
+{
+    public class ActionConstraintCache
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Core/Internal/ActionConstraintCache.cs]: Would we have a flag that would ignore missing doc comments when we set `xmlDocs: true`? If not, consider adding some docs here.

Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.Core/Internal/ActionConstraintCache.cs]: Yes. We have suppress error support.

**Comment by rynowak** [comment]
:shipit:

File: src/Microsoft.AspNet.Mvc.Core/Internal/ActionConstraintCache.cs [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        {
+            var cache = CurrentCache;
+
+            CacheEntry entry;
+            if (cache.Entries.TryGetValue(action, out entry))
+            {
+                return GetActionConstraintsFromEntry(entry, httpContext, action);
+            }
+
+            if (action.ActionConstraints == null || action.ActionConstraints.Count == 0)
```
Comment (javiercn) [inline src/Microsoft.AspNet.Mvc.Core/Internal/ActionConstraintCache.cs]: ```action.ActionConstraints?..Count == 0```?

File: src/Microsoft.AspNet.Mvc.Core/Internal/ActionConstraintCache.cs [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+                {
+                    current = new InnerCache(actionDescriptors.Version);
+                    _currentCache = current;
+                }
+
+                return current;
+            }
+        }
+
+        public IReadOnlyList<IActionConstraint> GetActionConstraints(HttpContext httpContext, ActionDescriptor action)
```
Comment (javiercn) [inline src/Microsoft.AspNet.Mvc.Core/Internal/ActionConstraintCache.cs]: null check arguments

**Pull request merged**